pkg/controller/routestate/route_state_controller.go

```diff
@@ -0,0 +1,246 @@
+func (c *RouteStateController) checkPodStatus(reference *corev1.ObjectReference) []string {
+	pod, err := c.podGetter.Pods(reference.Namespace).Get(reference.Name, metav1.GetOptions{})
+	if err != nil {
+		return []string{fmt.Sprintf("error getting pod %q: %v", reference.Name, err)}
+	}
+	result := []string{}
+	for _, containerStatus := range pod.Status.ContainerStatuses {
+		if containerStatus.Ready && containerStatus.State.Running != nil {
+			continue
+		}
+		if terminated := containerStatus.State.Terminated; terminated != nil {
+			result = append(result, fmt.Sprintf("pod %q container %q terminated with %q", pod.Name, containerStatus.Name, terminated.Message))
+		}
+		if containerStatus.RestartCount > 1 {
+			result = append(result, fmt.Sprintf("pod %q container %q restarted %d times", pod.Name, containerStatus.Name, containerStatus.RestartCount))
+		}
+		// TODO: Add more here, like we can detect "pending" pods
+	}
+	return result
+}
+
+// checkEndpointHealthz will check the health of given https://endpointIP:6443/healthz
+// TODO: this is insecure for now, but enough to tell us if we can get through
+func (c *RouteStateController) checkEndpointHealthz(endpointIP string) (bool, error) {
+	// TODO: consider making this secure, however it will mean wiring the trust which must match the same trust as ingress controller
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	response, err := http.Get(fmt.Sprintf("https://%s:6443/healthz", endpointIP))
+	if err != nil {
+		reportedError := err
+		if response != nil {
+			reportedError = fmt.Errorf("status:%q, body: %q, error: %v", response.Status, response.Body, reportedError)
+		}
+		return false, reportedError
+	}
+	if response.StatusCode != http.StatusOK {
+		return false, fmt.Errorf("status:%q, body: %q", response.Status, response.Body)
+	}
+	return true, nil
+}
+
+// degradedConditionNames lists all conditions that this controller manage
+var degradedConditionNames = []string{
+	"RouteStateEndpointsDegraded",
+	"RouteStatePodsDegraded",
+}
+
+func (c *RouteStateController) sync() error {
+	foundConditions := []operatorv1.OperatorCondition{}
+
+	endpoint, err := c.endpointsGetter.Endpoints(c.targetNamespace).Get("oauth-openshift", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if len(endpoint.Subsets) == 0 {
+		foundConditions = append(foundConditions, operatorv1.OperatorCondition{
+			Type:    "RouteStateEndpointsDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "ZeroEndpointsSubsets",
+			Message: "No endpoints found for oauth-server",
+		})
+	}
+
+	for _, subset := range endpoint.Subsets {
+		// If endpoint does not have any ready addresses, only not-ready, report it as RouteStateEndpointsDegraded
+		if len(subset.Addresses) == 0 && len(subset.NotReadyAddresses) > 0 {
+			foundConditions = append(foundConditions, operatorv1.OperatorCondition{
+				Type:    "RouteStateEndpointsDegraded",
```

Comment:
Won't this report `Degraded` during install?

Reply:
sure, but it should be just minor blip.

Reply:
I think there used to be a test that was checking the presence of degraded during installation. Would you be fine moving this to Progressing instead?